File: libraries/cyclestreets-view/src/main/java/net/cyclestreets/views/overlay/LockScreenOnOverlay.java
package net.cyclestreets.views.overlay;

import android.widget.Toast;
import net.cyclestreets.util.Theme;
import net.cyclestreets.view.R;
import net.cyclestreets.views.CycleMapView;

import org.osmdroid.views.MapView;

import android.content.Context;
import android.content.SharedPreferences;
import android.content.SharedPreferences.Editor;
import android.graphics.Canvas;
import android.graphics.drawable.Drawable;
import android.support.design.widget.FloatingActionButton;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;

import com.mikepenz.google_material_typeface_library.GoogleMaterial;
import com.mikepenz.iconics.IconicsDrawable;

import org.osmdroid.views.overlay.Overlay;

public class LockScreenOnOverlay extends Overlay implements PauseResumeListener {
  private static String LOCK_PREF = "lockScreen";

  private final CycleMapView mapView;

  private final FloatingActionButton screenLockButton;
  private final Drawable onIcon;
  private final Drawable offIcon;

  public LockScreenOnOverlay(final Context context, final CycleMapView mapView) {
    super();
    this.mapView = mapView;

    onIcon = new IconicsDrawable(context)
        .icon(GoogleMaterial.Icon.gmd_lock_open)
        .color(Theme.highlightColor(context))
        .sizeDp(24);
    offIcon = new IconicsDrawable(context)
        .icon(GoogleMaterial.Icon.gmd_lock_open)
        .color(Theme.lowlightColor(context))
        .sizeDp(24);

    View liverideButtonView = LayoutInflater.from(mapView.getContext()).inflate(R.layout.liveride_buttons, null);
    screenLockButton = liverideButtonView.findViewById(R.id.liveride_screenlock_button);
    screenLockButton.setOnClickListener(view -> screenLockButtonTapped());
    mapView.addView(liverideButtonView);

    mapView.setKeepScreenOn(false);
  }

  private void screenLockButtonTapped() {
    setScreenLockState(!mapView.getKeepScreenOn());
  }

  private void setScreenLockState(boolean state) {
    Log.d("LiveRide", "Setting keepScreenOn state to " + state);
    screenLockButton.setImageDrawable(state ? onIcon : offIcon);
    int message = state ? R.string.liveride_keep_screen_on_enabled : R.string.liveride_keep_screen_on_disabled;
    Toast.makeText(mapView.getContext(), message, Toast.LENGTH_LONG).show();
    mapView.setKeepScreenOn(state);
  }

  @Override
  public void draw(Canvas c, MapView osmv, boolean shadow) {}

  /////////////////////////////////////////
  @Override
  public void onResume(final SharedPreferences prefs) {
    mapView.setKeepScreenOn(prefs.getBoolean(LOCK_PREF, false));
  }

  @Override
  public void onPause(final Editor prefs) {
    prefs.putBoolean(LOCK_PREF, mapView.getKeepScreenOn());
  }
}

File: libraries/cyclestreets-view/src/main/java/net/cyclestreets/views/overlay/LockScreenOnOverlay.kt
package net.cyclestreets.views.overlay

import android.content.SharedPreferences
import android.graphics.Canvas
import android.graphics.drawable.Drawable
import android.util.Log
import android.view.LayoutInflater
import android.widget.Toast
import com.google.android.material.floatingactionbutton.FloatingActionButton
import com.mikepenz.iconics.typeface.library.googlematerial.GoogleMaterial
import net.cyclestreets.iconics.IconicsHelper.materialIcon
import net.cyclestreets.util.Theme.highlightColor
import net.cyclestreets.util.Theme.lowlightColor
import net.cyclestreets.view.R
import net.cyclestreets.views.CycleMapView
import org.osmdroid.views.MapView
import org.osmdroid.views.overlay.Overlay


class LockScreenOnOverlay(private val mapView: CycleMapView) : Overlay(), PauseResumeListener {

    companion object {
        private const val LOCK_PREF = "lockScreen"
    }

    private val screenLockButton: FloatingActionButton
    private val onIcon: Drawable
    private val offIcon: Drawable

    init {
        val context = mapView.context

        onIcon = materialIcon(context, GoogleMaterial.Icon.gmd_phonelink_lock, highlightColor(context))
        offIcon = materialIcon(context, GoogleMaterial.Icon.gmd_phonelink_lock, lowlightColor(context))

        val liveRideButtonView = LayoutInflater.from(context).inflate(R.layout.liveride_buttons, null)
        screenLockButton = liveRideButtonView.findViewById<FloatingActionButton>(R.id.liveride_screenlock_button).apply {
            setOnClickListener { _ -> screenLockButtonTapped() }
            setImageDrawable(offIcon)
        }
        mapView.addView(liveRideButtonView)
        mapView.keepScreenOn = false
    }

    private fun screenLockButtonTapped() {
        setScreenLockState(!mapView.keepScreenOn)
    }

    private fun setScreenLockState(state: Boolean) {
        Log.d("LiveRide", "Setting keepScreenOn state to $state")
        screenLockButton.setImageDrawable(if (state) onIcon else offIcon)
        val message = if (state) R.string.liveride_keep_screen_on_enabled else R.string.liveride_keep_screen_on_disabled
        Toast.makeText(mapView.context, message, Toast.LENGTH_LONG).show()
        mapView.keepScreenOn = state
    }

    override fun draw(c: Canvas, osmv: MapView, shadow: Boolean) {}

    /////////////////////////////////////////
    override fun onResume(prefs: SharedPreferences) {
        mapView.keepScreenOn = prefs.getBoolean(LOCK_PREF, false)
    }

    override fun onPause(prefs: SharedPreferences.Editor) {
        prefs.putBoolean(LOCK_PREF, mapView.keepScreenOn)
    }

}
